src/Streamly/Internal/Data/Stream/StreamD/Type.hs

```diff
@@ -231,23 +232,61 @@ concatMap f = concatMapM (return . f)
 
 -- | Create a singleton 'Stream' from a pure value.
 {-# INLINE_NORMAL yield #-}
-yield :: Monad m => a -> Stream m a
-yield x = Stream (\_ s -> return $ step undefined s) True
+yield :: Applicative m => a -> Stream m a
+yield x = Stream (\_ s -> pure $ step undefined s) True
   where
     {-# INLINE_LATE step #-}
     step _ True  = Yield x False
     step _ False = Stop
 
-instance Monad m => Applicative (Stream m) where
+{-# INLINE_NORMAL concatAp #-}
+concatAp :: Functor f => Stream f (a -> b) -> Stream f a -> Stream f b
+concatAp (Stream stepa statea) (Stream stepb stateb) = Stream step' (Left statea)
```

Comment:
Instead of stepa/statea I generally use step1/state1, that way it is more readable.